### PR TITLE
Improve ClientConnectionProcessListenerTest [API-1998]

### PR DESCRIPTION
**Pull request by mdumandag**
The test failed with logs indicating that one member we were expecting to start in port 5701 started in port 5702, which resulted in failed assertions.

To fix this and some other problems, I have refactored to tests a bit.

- Now, we don't depend on any specific port and perform our assertions based on the actual port of the member.
- I have set the cluster connect timeout to 0, so that the tests we expect the client to fail to connect can complete faster.
- Improved the read-handling code in the `remoteClosesConnection` test to make sure that the test would pass reliably.
- Removed the unneeded server instance from `hostResolutionFailure_portFailure` test.
- Used client selectors instead of relying on some implementation details of the failover client in `clientNotAllowedInCluster` test.

closes https://github.com/hazelcast/hazelcast/issues/23520

### CODE REVIEW

File: hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java [hunk 1]
```diff
@@ -66,168 +67,169 @@ private static ClientConfig newClientConfig() {
...
         assertTrueEventually(() -> {
-            verify(listener).possibleAddressesCollected(asList(
-                    new Address("127.0.0.1", 5701),
-                    new Address("127.0.0.1", 5702),
-                    new Address("127.0.0.1", 5703)
-            ));
-            verify(listener, atLeastOnce()).attemptingToConnectToAddress(new Address("127.0.0.1", 5701));
-            verify(listener).authenticationSuccess(new Address("127.0.0.1", 5701));
-            verify(listener).clusterConnectionSucceeded(clusterName);
+            verify(listener).possibleAddressesCollected(singletonList(address));
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java:92]: It is not members, it is just candidate member addresses. If you just give an address without a port, we will try to connect to addr:5701, addr:5702, and addr:5703. When you don't give any addresses, we add `localhost` to the candidate address list, so we were "collecting" 3 candidate member addresses

File: hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java [hunk 2]
```diff
@@ -66,168 +67,169 @@ private static ClientConfig newClientConfig() {
...
+            verify(listener, atLeastOnce()).attemptingToConnectToAddress(address);
+            verify(listener).authenticationSuccess(address);
+            verify(listener).clusterConnectionSucceeded(getTestMethodName());
             verifyNoMoreInteractions(listener);
         });
     }
 
     @Test
     public void hostResolutionFailure_portFailure() throws Exception {
```
[the remark this answers is not in the document]
Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java:101]: How is it different? 

- You have a member on 5701 and try to connect to only 5702 
- You don't have a member on 5701 and try to connect to only 5702

From the client's PoV, you cannot find a member at the address you try to connect in both of the cases.

File: hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java [hunk 3]
```diff
@@ -66,168 +67,169 @@ private static ClientConfig newClientConfig() {
...
-        Hazelcast.newHazelcastInstance(new Config().setClusterName(clusterName));
+        ClientConnectionProcessListener listener = mock(ClientConnectionProcessListener.class);
+        ClientConfig clientConfig = newClientConfig();
+        clientConfig.addListenerConfig(new ListenerConfig(listener))
+                .getNetworkConfig()
+                .addAddress(address.getHost() + ":" + address.getPort());
 
         try {
-            HazelcastClient.newHazelcastFailoverClient(failoverConfig);
+            HazelcastClient.newHazelcastClient(clientConfig);
```
Comment (emreyigit) [inline hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java:191]: Why did you change it to normal client?

Comment (mdumandag) [reply hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java:191]: To make it simpler. The previous version of the test relied on some implementation details about when we threw that particular exception. With client selectors, it is more intuitive

**Comment by mdumandag** [comment]
@emreyigit @srknzl I have answered your question, can you take another look when you have the time, please?

**Review by emreyigit — APPROVED** [review]

File: hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (srknzl) [reply hazelcast/src/test/java/com/hazelcast/client/impl/management/ClientConnectionProcessListenerTest.java:101]: ok. makes sense.

**Review by srknzl — APPROVED** [review]
